### PR TITLE
TIG-3498 Support multiple tenants in Genny

### PR DESCRIPTION
**Pull request by aggrand**
This treats the URI as part of the client pool configuration. It adds a preprocessing step to inject/configure the default pool, and treats `--mongo-uri` as a syntactic sugar preprocessing step to set the default URI for all pools. Users needing to connect to multiple tenants can use the override config syntax to set them at runtime.

Adding more preprocessing is tricky, but I think python has better tools for managing configuration complexity than C++ does. I'd like to keep the C++ config as straightforward and "dumb" as possible, just going off the given yaml and doing all config management in python. I know allowing arbitrary overrides gives users enough rope to hang themselves, but I think in the future we can use OmegaConf's structured configs to perform validation and allow different components to be configured together with minimal coupling. (Probably something for a future project though.)

As a side effect of this change, workloads will now be run as `./run-genny workload -u <URI> <YAML>`.

DSI change: https://github.com/10gen/dsi/pull/1007

### CODE REVIEW

File: src/cast_core/src/Loader.cpp
Comment (aggrand) [inline src/cast_core/src/Loader.cpp:198]: I didn't realize at first that the `Client` option was a `CrudActor` internal thing. Would it be worth adding this option to all actors?

**Comment by aggrand** [comment]
cc @rtimmons in case you have thoughts on the config placement / syntax! Wasn't sure if you'd feel strongly about putting this in preprocessing.

cc @visualzhou @JamesHeppenstall to see if the syntax works for y'all when adding the multitenant test fixture and piping the URI through. The upshot is to create multiple client pools, distribute them among the actors, pass in an override file when invoking genny with `-o`, and have the URIs in the override file at the key `Clients/<PoolName>/URI/<URIValue>`.

Tested locally by spawning two mongods and running a workload against both.

File: src/cast_core/src/Loader.cpp
Comment (vrachev) [reply src/cast_core/src/Loader.cpp:198]: Looks like runCommand has it as well. I'm assuming people add this to their actor as they see the need. Up to you on whether to add or not, but would be a nice change

**Review by vrachev — CHANGES_REQUESTED** [review]
Nice work, and I like the addition of OmegaConf even if we'll need to spend a little more time integrating it. 

I thought a little about the best way to input URIs, and I think I agree that an override yaml file is the cleanest way. It's better than having the URI logic in DSI and cleaner than passing in a list of strings to the CLI. Still it does worry me a little to have yet another source of unbounded config. Can you add a doc guiding users on what config should go in the override file vs the workload yaml (and auto_task section)? And perhaps we should file a ticket to document the config story across the genny/dsi/mongodb repos, although we'd hopefully simplify it when we turn these into services. 

I think it would also be nice to add a doc explaining the multi-tenant use-case and our support for it, with a small example workload + override file. The workload could live in /docs or something, up to you.

I had one small nit as well in a comment. Thanks for the work on this!

**Review by rtimmons — COMMENTED** [review]
Overall the code seems reasonable, although I think more real-world test-cases that include invalid or perhaps quirky override situations would be helpful.

This said: this change makes me uneasy about the future of Genny config. This could just be me being out of the headspace for a while though.

An alternative would have been to simply allow a syntax that the preprocessor uses to resolve parameters from `mongodb_setup.yml`:


```yaml
Clients:
  Default:
    URI:
      ^Parameter:
        FromYaml: mongodb_setup.foo.bar.uri
```

This is much less powerful than what you have, but it is conceptually much simpler and less reality-shattering.

(We could also rely on DSI to parse ConfigDict values `URI: {^FromDsi: "${setup.foo}"}` which is resolved via `run-dsi get-config setup.foo` or something. But this may be imbibing too much ConfigDict.)

The decision on which approach is "better" is subjective and also utlimately not super far-reaching (yet), so I will leave it to you to decide.

My input would mostly be let's not shy away from re-evaluating what parts of Genny's syntax still make sense and what we can axe. If we were to write the syntax from scratch today I doubt we'd land all that close to what we have. 

The syntax was conceived 3 years ago before we had parameters, deeper nop support, includes, files across repos, and a bevy of examples of workload yamls to compare before/after syntax changes on. It started off with a cascading notion where `X.Y` would resolve to `X` unless present. Hence `Operation:OperationName` rather than `Operation:Name`. This was quickly deemed fraught and abandoned but the premise of the design wasn't reconsidered. The syntax was also designed to be somewhat straightforward to parse in C++ which is no longer a deep requirement.

The overrides concepts in this PR may be papering over a existing design limitation rather than being a uniquely necessary bit of kit.

A syntax re-think probably isn't as arduous as it sounds. Nearly all workload yamls live in the genny repo, and `--evaluate` or whatever makes for a fast iteration cycle. We delay this at our peril: The more users add yamls to private repos (and the more yamls in genny's own repo) the harder it becomes since back-compat is an annoyance.

Again, this is an idea that I will let you pursue as you see fit. Of course I'm happy to collaborate further if it would be helpful. As it stands, I'll give my usual godspeed and see how this pans out.

File: src/cast_core/src/Loader.cpp
Comment (aggrand) [reply src/cast_core/src/Loader.cpp:198]: Cool created TIG-3714 for this.

File: src/gennylib/src/PoolManager.cpp
Comment (aggrand) [reply src/gennylib/src/PoolManager.cpp:26]: I'm kinda hesitant to default both here and in the preprocessor. I feel like I'd like to get to a point where the C++ code doesn't do any defaulting so we don't have to check several places, and what's configured can be more easily seen from the `evaluate` command. That said, push back if you disagree!

File: src/lamplib/src/genny/cli.py
Comment (aggrand) [reply src/lamplib/src/genny/cli.py:269]: Nope it's removed and not necessary! I kinda feel like the options to the core binary are just internal implementation details of the python CLI.

File: src/lamplib/src/genny/tasks/preprocess.py
Comment (aggrand) [reply src/lamplib/src/genny/tasks/preprocess.py:58]: Ooh good point. I checked and it does not! Looks like omegaconf deep-copies the dict.

File: src/lamplib/requirements.txt
Comment (aggrand) [reply src/lamplib/requirements.txt:7]: Good point, added overrides stuff to the `Using` doc and added a connection-pool-specific one.

File: src/lamplib/src/tests/test_preprocess.py
Comment (aggrand) [reply src/lamplib/src/tests/test_preprocess.py:439]: Good point, done!

File: src/workloads/scale/LargeIndexedIns.yml
Comment (aggrand) [reply src/workloads/scale/LargeIndexedIns.yml:20]: The CrudActor in this workload had tried to retrieve a client pool named `Update`, but it didn't exist. A side effect of this change is that any undeclared pools cause an error at dry-run evaluation. So I added this one, shouldn't change behavior since `maxPoolSize` defaults to 100 anyway and before it was creating an unconfigured default pool named `Update`. Will confirm in a patch before merging.

Comment (aggrand) [reply src/workloads/scale/LargeIndexedIns.yml:20]: Will document the need to declare pools explicitly. That behavior has pros and cons, but I'm honestly not sure in this case whether it was intentional or not, so forced explicitness might prevent footgunning.

File: src/lamplib/src/genny/tasks/preprocess.py
Comment (aggrand) [reply src/lamplib/src/genny/tasks/preprocess.py:65]: That's a really good point. I'm going to capture this and your thoughts in the top-level comment in a separate ticket. My instinct is to move forward as-is but definitely do a more thorough evaluation of Genny's configuration scheme. In particular, I have half-baked ideas of using OmegaConf structured configs for each cpp module. This could let us distribute the config schemas more while type validation lets us do that safely.

The `SchemaVersion` stuff in particular ties into the infra versioning concerns, which we have an (admittedly unstarted) initiative to do and a Genny-specific project for. I especially want to unify Genny and DSI configs. It'd be really nice to have a global `evaluate` before launching a DSI run that would merge in the Genny config too and let you tweak/override things. Will defs keep you in the loop I think this needs a lot more conversation.

Comment (aggrand) [reply src/lamplib/src/genny/tasks/preprocess.py:65]: Created TIG-3717 for this.

File: src/lamplib/src/genny/cli.py
Comment (rtimmons) [reply src/lamplib/src/genny/cli.py:282]: > the options to the core binary are just internal implementation details

Ya very good point. Now that we've got the `debug`/`hang` thing plus the curator handler there's very little reason for anyone to interact with `genny_core` directly. Let's hide it

(Maybe even consider setting a `GENNY_RUNNING_FROM_PYTHON` env var in the python (but with a non-terrible name). Then `genny_core` looks for that env var and barfs or yells if it doesn't see it. That may be too cheeky, but it would set the expectation correctly.)

Resolving the convo. This is good how it is.

File: src/lamplib/src/genny/tasks/preprocess.py
Comment (rtimmons) [reply src/lamplib/src/genny/tasks/preprocess.py:72]: Ya definitely an interesting idea to start and bike-shed over a lunch or happy hour until it starts to feel real and do-able. I'll look forward to that.

**Comment by aggrand** [comment]
Sys-perf patch on standalone: https://spruce.mongodb.com/version/61db279f2a60ed121efed998/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

**Review by vrachev — APPROVED** [review]
Nice! This will be cool to see being used

**Comment by brooookemiller** [comment]
Ping @rtimmons please take a look when you can 😃  I presume @aggrand is waiting for your approval

**Comment by aggrand** [comment]
@brooookemiller ryan didn't request any changes, so I wasn't blocking there (though Ryan any more thoughts are always welcome!) I was waiting on the DSI-side PR to be approved, which has happened now.

**Comment by aggrand** [comment]
evergreen merge

**Comment by aggrand** [comment]
Squash merging to align with the required DSI squash merge (to minimize the interval between commits).

**Pull request merged**
